### PR TITLE
Support ArrayBuffer in BodyInit

### LINKED ISSUE #350: BodyInit extract is not supported with arrayBuffer
~~https://github.com/github/fetch/commit/1529c832406dd7f3e78fe6670fae3a29fcdf8c3c added support for arrayBuffer() for POST responses only, but I can't tell why.~~

~~Specifically, in https://github.com/cockroachdb/cockroach/pull/7178 I am relying on arrayBuffer() working in GET responses, which fails under whatwg-fetch but works in Chrome because chrome has native `fetch` support.~~

~~Was there a reason that arrayBuffer() support was omitted for non-POST responses? If not, could you add this support now?~~

EDIT: turns out this is just a simple issue with BodyInit; see https://fetch.spec.whatwg.org/#concept-bodyinit-extract.

### PR DESCRIPTION
**Pull request by tamird**
Fixes #350.

### CODE REVIEW

**Comment by tamird** [comment]
Sauce build failed, but the main build is green.

**Comment by mislav** [comment]
Don't worry about Sauce failures for now. But, you should definitely test this manually in browsers that don't support native `fetch()` (such as Safari and IE).

Thanks for the fix. What was the original problem, in your opinion. And, is it possible to simplify the fix? Because, the code seems right now it's doing a lot to just store an ArrayBuffer for body. Where does the complexity come from?

Thanks!

**Comment by tamird** [comment]
Indeed, I've tested this in safari as well as phantomjs.

Regarding the fix, there are two bits of complexity:
1. the spec says to copy the incoming buffer. Ideally, one would use `.slice(0)` for this, but the version of phantomjs that this repo's tests run against doesn't support that method, hence the `if (body.slice)` check and alternative implementation.
2. internally, this library feeds a blob into `Response`'s constructor, so the old way of reading `arrayBuffer()` from `_bodyBlob` must still be supported.

I'm open to suggestions on how to simplify this fix.

**Comment by mislav** [comment]
Should we treat browsers without `ArrayBuffer.slice` support as not supporting ArrayBuffer altogether? Then we can skip the related tests. No sense in providing an alternative implementation that only helps PhantomJS.

**Comment by tamird** [comment]
I don't think the issue is specific to phantomjs, I would expect all older
WebKit browsers to be affected. Why not provide this support? It's pretty
simple to do so.

File: fetch.js [hunk 1]
```diff
@@ -236,6 +236,21 @@
       }
     }
 
+    if (support.arrayBuffer) {
+      this.arrayBuffer = function() {
+        if (this._bodyArrayBuffer) {
+          var rejected = consumed(this)
+          if (rejected) {
+            return rejected
```
Comment (mislav) [inline fetch.js]: What do you think about condensing this whole condition like:

```js
return consumed(this) || Promise.resolve(...)
```

Comment (tamird) [inline fetch.js]: I'm fine with this refactor, but this pattern is everywhere in this file. Doesn't seem appropriate to hold this change up over that style debate.

File: fetch.js [hunk 2]
```diff
@@ -236,6 +236,21 @@
...
+          }
+
+          return Promise.resolve(this._bodyArrayBuffer.buffer)
+        } else if (support.blob) {
```
Comment (mislav) [inline fetch.js]: I wouldn't test for Blob support here. First of all, do any browsers implement ArrayBuffer but not Blob? Second, I don't want this function to ever return `undefined`.

**Comment by mislav** [comment]
Thanks for the explanation. Have you looked into #343 and #344? Do you think those concerns should affect this PR? /cc @datenreisender

File: fetch.js (hunk 1 above)
Comment (mislav) [inline fetch.js]: It's fine to not match the style of others in this instance. Other invocations of `consumed()` can not efficiently be condensed into a one-liner — except for `text()`, which already uses a ternary operator.

File: fetch.js (hunk 2 above)
Comment (tamird) [inline fetch.js]: Indeed, it seems that `ArrayBuffer` is more widely supported than `Blob`. Fixed this.

File: fetch.js (hunk 1 above)
Comment (tamird) [inline fetch.js]: Done.

**Comment by tamird** [comment]
#343 isn't directly related - fixing it requires wrapping before the xhr. #344 is a bit more closely related, but @datenreisender's last suggestion in that thread is already the correct fix - this PR doesn't make that any more difficult.

**Comment by tamird** [comment]
@mislav @datenreisender I've added a commit to fix #343.

**Comment by tamird** [comment]
Sigh, can't reproduce this CI failure locally.

File: fetch.js [hunk 3]
```diff
@@ -426,7 +435,11 @@
         xhr.setRequestHeader(name, value)
       })
 
-      xhr.send(typeof request._bodyInit === 'undefined' ? null : request._bodyInit)
+      if (request._bodyArrayBuffer) {
+        xhr.send(request._bodyArrayBuffer)
```
Comment (mislav) [inline fetch.js]: Shouldn't we be avoiding passing ArrayBuffer into xhr directly?

Comment (mislav) [inline fetch.js]: Also, it would be good to make a test for this that performs a real HTTP request (in supporting browsers).

Comment (tamird) [inline fetch.js]: The tests already exists, or are you suggesting that I test for a deprecation warning? I don't know how to do that.

Comment (tamird) [inline fetch.js]: The type of `_bodyArrayBuffer` is actually `ArrayBufferView`

Comment (mislav) [inline fetch.js]: Ah I see. That clears it up; thanks!

**Comment by tamird** [comment]
Updated to try to solve the second commit's test failure.

**Comment by tamird** [comment]
Removed the last commit - passing ArrayBufferView into xhr results in the literal `'[object Uint8Array]'` being sent over the wire.

The version of phantomjs used in CI here is just too old.

**Comment by tamird** [comment]
@mislav I've taken this as far as I can. Will you put me out of my misery? 👼

**Comment by mislav** [comment]
You've helped enough and I thank you. We'll take it over from here. I'll have to look into some things first.

**Comment by tamird** [comment]
@mislav any news here?

**Comment by mislav** [comment]
No, sorry! I need a bit more time.

**Comment by gaye** [comment]
+1

**Comment by tamird** [comment]
@mislav any news?

**Comment by mislav** [comment]
Thanks for rebasing! Haven't gotten around to this yet. Patience!

**Comment by tmbo** [comment]
@mislav any progress here? +1

**Comment by mislav** [comment]
@tamird I want to proceed with this PR, but I need help from you to clarify something.

As far as I understand, these APIs are currently broken in the polyfill, before your changes:

1. A Request instance can be initialized with an ArrayBuffer body, but that body is then not readable via `request.arrayBuffer()` method.
2. Same as (1), but for Response.

Although I agree that these inconsistencies are problematic and need to be fixed, I still don't understand how they interfered with sending & receiving of binary data. These are the facts right now:

1. The `fetch` method **doesn't use** the `request.arrayBuffer()` method to obtain the ArrayBuffer body object. Instead, it reads it directly from `request._bodyInit`. Therefore, sending ArrayBuffers is already possible, and there is a test for that.

2. The `fetch` method **doesn't ever initialize** a Response object with an ArrayBuffer body, because all responses returned from XHR are of type Blob. Therefore, receiving ArrayBuffers from response is already possible, because Blob → ArrayBuffer conversion is functional and achieved through `response.arrayBuffer()`, and there is a test for that too.

Can you help me construct a simple test case that utilizes the `fetch` method to communicate with ArrayBuffers and fails without your fixes? That would be a good starting point for me to get this merged.

**Comment by tamird** [comment]
A great way to ensure that I don't remember the detail is of this is so wait 5 months before asking questions.

We're still using my fork of this library, so it's not as if this has become irrelevant, I just can't remember the details.

> The fetch method doesn't use the request.arrayBuffer() method to obtain the ArrayBuffer body object. Instead, it reads it directly from request._bodyInit. Therefore, sending ArrayBuffers is already possible, and there is a test for that.

This seems patently incorrect based on #350 and the new tests added in this PR. Where is the test you claim exists?

Not sure about your second bullet.

**Comment by mislav** [comment]
> This seems patently incorrect based on #350 and the new tests added in this PR. Where is the test you claim exists?

[It's here](https://github.com/github/fetch/blob/87fd1cd3ad6651ef34e8ba7b667dc19ce0ec865f/test/test.js#L871-L881). It passes even under PhantomJS.

Apologies for sitting on this for so long, but you have to consider my perspective as a maintainer: I have received a pull request that I'm not sure what it solves, has failing tests in CI, and is for a feature that we ourselves don't use at GitHub. This wasn't a really high priority for me.

I'm sure that there is a valid bug somewhere that you discovered, especially since on the cocroachdb project you reported that tests fail on PhantomJS and Safari, which are the browsers where our polyfill is in effect. I just need to first discover how this bug affects `fetch` calls that operate on binary data. I'm working on a more complete ArrayBuffer support, starting with #430.

**Comment by tamird** [comment]
Heh, the failing tests were originally due to sauce labs failing, and after today's rebase were due to a change in the Request constructor, so please, don't try to misrepresent. The "fix" was:
```diff
diff --git a/test/test.js b/test/test.js
index a066e8d..fed12b4 100644
--- a/test/test.js
+++ b/test/test.js
@@ -455,7 +455,7 @@ suite('Request', function() {
           view[i] = text.charCodeAt(i)
         }
 
-        var request = new Request(null, {method: 'POST', body: buf})
+        var request = new Request('', {method: 'POST', body: buf})
         return request.arrayBuffer().then(function(buffer) {
           var bufView = new Uint8Array(buffer)
           assert.equal(bufView.byteLength, view.length)

```

And was needed due to a change you made long after this PR was opened 5f61b12416bc4497f23f67da562b4219b54b49c2

**Comment by tamird** [comment]
I've just confirmed again that moving from my fork to mainline causes tests in CRDB to fail. For what it's worth, here's where we're using this: https://github.com/cockroachdb/cockroach/blob/428c303/pkg/ui/app/util/api.ts#L90:L113

**Comment by tamird** [comment]
Ah! I remember now, this doesn't actually have anything to do with `fetch` itself, it has to do with our use of `fetch-mock` where we need to manually initialize the response object like we do here: https://github.com/cockroachdb/cockroach/blob/428c30331c2ab153450afa110cba6e119e990ec1/pkg/ui/app/util/api.spec.ts#L109:L113

**Comment by mislav** [comment]
> Heh, the failing tests were originally due to sauce labs failing, and after today's rebase were due to a change in the Request constructor, so please, don't try to misrepresent.

Apologies. 🙇 

> I remember now, this doesn't actually have anything to do with `fetch` itself, it has to do with our use of `fetch-mock` where we need to manually initialize the response object

That makes perfect sense. Thanks for looking that up!

**Comment by tamird** [comment]
@mislav where does that leave us? Shall I rebase this again?

**Comment by mislav** [comment]
@tamird No, just leave it to me now. Thanks for the help so far!

**Comment by tamird** [comment]
🎊

**Comment by mislav** [comment]
@tamird Try out the new release and let me know if there are any more problems!

**Comment by tamird** [comment]
We're already using it in CRDB! Thanks again.

**Pull request merged**